src/sentry/tagstore/snuba/backend.py

```diff
@@ -0,0 +1,104 @@
+"""
+sentry.tagstore.snuba.backend
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+:copyright: (c) 2010-2018 by the Sentry Team, see AUTHORS for more details.
+:license: BSD, see LICENSE for more details.
+"""
+
+from __future__ import absolute_import
+
+from datetime import datetime, timedelta
+import pytz
+import six
+
+from sentry.tagstore import TagKeyStatus
+from sentry.tagstore.base import TagStorage
+from sentry.utils import snuba
+
+
+class SnubaTagStorage(TagStorage):
+
+    # Tag keys and values
+    def get_tag_key(self, project_id, environment_id, key, status=TagKeyStatus.VISIBLE):
+        pass
+
+    def get_tag_keys(self, project_id, environment_id, status=TagKeyStatus.VISIBLE):
+        pass
+
+    def get_tag_value(self, project_id, environment_id, key, value):
+        pass
+
+    def get_tag_values(self, project_id, environment_id, key):
+        pass
+
+    def get_group_tag_key(self, project_id, group_id, environment_id, key):
+        pass
+
+    def get_group_tag_keys(self, project_id, group_id, environment_id, limit=None):
+        pass
+
+    def get_group_tag_value(self, project_id, group_id, environment_id, key, value):
+        pass
+
+    def get_group_tag_values(self, project_id, group_id, environment_id, key):
+        pass
+
+    def get_group_list_tag_value(self, project_id, group_id_list, environment_id, key, value):
+        pass
+
+    def get_group_tag_value_count(self, project_id, group_id, environment_id, key):
+        pass
+
+    def get_group_tag_values_for_users(self, event_users, limit=100):
+        pass
+
+    def get_top_group_tag_values(self, project_id, group_id, environment_id, key, limit=3):
+        pass
+
+    # Releases
+    def get_first_release(self, project_id, group_id):
+        pass
+
+    def get_last_release(self, project_id, group_id):
+        pass
+
+    def get_release_tags(self, project_ids, environment_id, versions):
+        pass
+
+    def get_group_event_ids(self, project_id, group_id, environment_id, tags):
+        pass
+
+    def get_group_ids_for_users(self, project_ids, event_users, limit=100):
+        pass
+
+    def get_groups_user_counts(self, project_id, group_ids, environment_id):
+        pass
+
+    # Search
+    def get_group_ids_for_search_filter(self, project_id, environment_id, tags):
+        from sentry.search.base import ANY, EMPTY
+        filters = {
+            'environment': [environment_id],
+            'project_id': [project_id],
+        }
+
+        conditions = []
+        for tag, val in six.iteritems(tags):
+            col = 'tags[{}]'.format(tag)
+            if val == ANY:
+                conditions.append((col, 'IS NOT NULL', None))
+            elif val == EMPTY:
```

Comment:
I'm struggling to find/remember what `EMPTY` and `ANY` really mean again, but the other tagstore implementations seem to do something very different than checking whether the tag is null?

```
        for k, v in tag_lookups:
            if v is EMPTY:
                return None
```

:confused:

Reply:
I guess the impl of this method isn't that important to the PR though, I just wanted to point that out.

Reply:
Probably could use @tkaemming here too because looking at the existing impls I'm more confused than not.

Reply:
IIRC, `EMPTY` is basically the equivalent of calling `queryset.none()` and only appears to be generated as part of this function: https://github.com/getsentry/sentry/blob/124493699517d1858aa81b76d37161c4b852ff7f/src/sentry/search/utils.py#L20-L33

This basically precludes the necessity for doing any of the rest of the search (since we only `AND` conditions together, hitting this branch means there can't possibly be any results) so in my opinion we should should remove this from tagstore if it's not too much trouble.

Reply:
>  we should should remove this from tagstore

Do you just mean something at a higher level up should return `None`, so tagstore never has to see `EMPTY`? Or something different?

Reply:
In the meantime I've just put in an early return if there's any `EMPTY` tags, but open to considering a refactor where tagstore doesn't see `EMPTY` at all.

Reply:
> Do you just mean something at a higher level up should return `None`, so tagstore never has to see `EMPTY`?

Yep.

> In the meantime I've just put in an early return if there's any `EMPTY` tags, but open to considering a refactor where tagstore doesn't see `EMPTY` at all.

Makes sense to me. 👍